File: examples/use-cases/ai-agents/python/ai_agent.py
"""
AI Agent with Route Prioritization and Tool Orchestration

This example demonstrates:
- Intelligent query routing and prioritization
- Deterministic tool calling with built-in orchestration
- State management and failure handling
- Safety constraints and guardrails
"""

import json
import time
from typing import Dict, List, Any, Optional
from dataclasses import dataclass
from enum import Enum

from hatchet_sdk import Hatchet, Context


class QueryPriority(Enum):
    URGENT = "urgent"
    HIGH = "high" 
    NORMAL = "normal"
    LOW = "low"


@dataclass
class CustomerQuery:
    query_id: str
    customer_id: str
    message: str
    priority: QueryPriority
    context: Dict[str, Any]


@dataclass
class ToolResult:
    tool_name: str
    success: bool
    result: Any
    execution_time: float


@dataclass
class AgentState:
    query_id: str
    current_step: str
    tool_results: List[ToolResult]
    conversation_history: List[Dict[str, str]]
    safety_violations: List[str]


hatchet = Hatchet(debug=True)


@hatchet.workflow(on_events=["customer.query"])
class AIAgentWorkflow:
    """
    AI Agent workflow with intelligent routing and tool orchestration
    """

    @hatchet.step(timeout="30s", retries=2)
    def route_and_prioritize(self, context: Context) -> dict:
        """Route and prioritize customer queries intelligently"""
        
        data = context.workflow_input()
        query = CustomerQuery(**data)
        
        # Intelligent routing based on query content and customer tier
        route_decision = {
            "query_id": query.query_id,
            "assigned_agent": self._determine_agent_type(query),
            "priority_score": self._calculate_priority_score(query),
            "estimated_complexity": self._assess_complexity(query.message),
            "required_tools": self._identify_required_tools(query.message)
        }
        
        print(f"Query {query.query_id} routed to {route_decision['assigned_agent']} with priority {route_decision['priority_score']}")
        
        return route_decision

    @hatchet.step(timeout="60s", retries=1, parents=["route_and_prioritize"])
    def execute_tools(self, context: Context) -> dict:
        """Execute deterministic tooling with orchestration primitives"""
        
        route_info = context.step_output("route_and_prioritize")
        required_tools = route_info["required_tools"]
        
        state = AgentState(
            query_id=route_info["query_id"],
            current_step="execute_tools",
            tool_results=[],
            conversation_history=[],
            safety_violations=[]
        )
        
        # Execute tools with safety constraints
        for tool_name in required_tools:
            try:
                # Safety check before tool execution
                if not self._check_safety_constraints(tool_name, context):
                    state.safety_violations.append(f"Safety violation prevented {tool_name} execution")
                    continue
                
                start_time = time.time()
                result = self._execute_tool(tool_name, context)
                execution_time = time.time() - start_time
                
                tool_result = ToolResult(
                    tool_name=tool_name,
                    success=True,
                    result=result,
                    execution_time=execution_time
                )
                state.tool_results.append(tool_result)
                
                print(f"Tool {tool_name} executed successfully in {execution_time:.2f}s")
                
            except Exception as e:
                print(f"Tool {tool_name} failed: {str(e)}")
                state.tool_results.append(ToolResult(
                    tool_name=tool_name,
                    success=False,
                    result=str(e),
                    execution_time=0
                ))
        
        return {
            "state": state.__dict__,
            "successful_tools": len([r for r in state.tool_results if r.success]),
            "failed_tools": len([r for r in state.tool_results if not r.success]),
            "safety_violations": len(state.safety_violations)
        }

    @hatchet.step(timeout="45s", parents=["execute_tools"])
    def generate_response(self, context: Context) -> dict:
        """Generate final response with state management"""
        
        tool_results = context.step_output("execute_tools")
        route_info = context.step_output("route_and_prioritize")
        
        # Check if human intervention is needed
        needs_human = (
            tool_results["failed_tools"] > 0 or
            tool_results["safety_violations"] > 0 or
            route_info["priority_score"] > 8
        )
        
        if needs_human:
            # Signal for Human-in-the-Loop
            response = {
                "status": "human_review_required",
                "reason": "Complex query or safety concerns detected",
                "query_id": route_info["query_id"],
                "escalation_priority": "high" if tool_results["safety_violations"] > 0 else "normal"
            }
        else:
            # Generate automated response
            response = {
                "status": "completed",
                "query_id": route_info["query_id"],
                "response": self._generate_final_response(tool_results, route_info),
                "confidence_score": self._calculate_confidence(tool_results),
                "processing_time": time.time()
            }
        
        print(f"Query {route_info['query_id']} processed with status: {response['status']}")
        
        return response

    def _determine_agent_type(self, query: CustomerQuery) -> str:
        """Determine the appropriate agent type based on query characteristics"""
        if "technical" in query.message.lower() or "api" in query.message.lower():
            return "technical_agent"
        elif "billing" in query.message.lower() or "payment" in query.message.lower():
            return "billing_agent"
        else:
            return "general_agent"

    def _calculate_priority_score(self, query: CustomerQuery) -> int:
        """Calculate priority score (1-10 scale)"""
        base_score = {
            QueryPriority.URGENT: 9,
            QueryPriority.HIGH: 7,
            QueryPriority.NORMAL: 5,
            QueryPriority.LOW: 3
        }[query.priority]
        
        # Adjust based on customer tier
        if query.context.get("customer_tier") == "enterprise":
            base_score += 2
        elif query.context.get("customer_tier") == "pro":
            base_score += 1
            
        return min(base_score, 10)

    def _assess_complexity(self, message: str) -> str:
        """Assess query complexity"""
        word_count = len(message.split())
        technical_terms = ["api", "integration", "webhook", "database", "authentication"]
        
        if word_count > 100 or any(term in message.lower() for term in technical_terms):
            return "high"
        elif word_count > 50:
            return "medium"
        else:
            return "low"

    def _identify_required_tools(self, message: str) -> List[str]:
        """Identify required tools based on message content"""
        tools = []
        
        if "status" in message.lower() or "health" in message.lower():
            tools.append("system_status_checker")
            
        if "account" in message.lower() or "billing" in message.lower():
            tools.append("account_lookup")
            
        if "documentation" in message.lower() or "how to" in message.lower():
            tools.append("knowledge_base_search")
            
        if "api" in message.lower():
            tools.append("api_validator")
            
        return tools if tools else ["general_assistant"]

    def _check_safety_constraints(self, tool_name: str, context: Context) -> bool:
        """Check safety constraints before tool execution"""
        
        # Example safety constraints
        dangerous_tools = ["system_admin", "database_modifier"]
        
        if tool_name in dangerous_tools:
            # Check if user has appropriate permissions
            user_role = context.workflow_input().get("context", {}).get("user_role", "user")
            if user_role not in ["admin", "superuser"]:
                return False
                
        # Rate limiting check
        if self._check_rate_limit(tool_name, context):
            return False
            
        return True

    def _check_rate_limit(self, tool_name: str, context: Context) -> bool:
        """Check if tool execution would exceed rate limits"""
        # Simplified rate limiting logic
        # In practice, this would check Redis or similar
        return False

    def _execute_tool(self, tool_name: str, context: Context) -> Any:
        """Execute the specified tool"""
        
        # Mock tool implementations
        tools = {
            "system_status_checker": lambda: {"status": "operational", "uptime": "99.9%"},
            "account_lookup": lambda: {"account_id": "12345", "tier": "pro", "status": "active"},
            "knowledge_base_search": lambda: {"results": ["Doc 1", "Doc 2"], "confidence": 0.85},
            "api_validator": lambda: {"valid": True, "last_tested": "2024-01-01"},
            "general_assistant": lambda: {"response": "I can help you with that!"}
        }
        
        if tool_name in tools:
            return tools[tool_name]()
        else:
            raise Exception(f"Unknown tool: {tool_name}")

    def _generate_final_response(self, tool_results: dict, route_info: dict) -> str:
        """Generate the final response based on tool results"""
        
        successful_tools = tool_results["successful_tools"]
        
        if successful_tools == 0:
            return "I apologize, but I encountered issues processing your request. Please try again or contact support."
        
        return f"Based on my analysis using {successful_tools} tools, I can help you with your query. Here's what I found..."

    def _calculate_confidence(self, tool_results: dict) -> float:
        """Calculate confidence score for the response"""
        
        total_tools = tool_results["successful_tools"] + tool_results["failed_tools"]
        if total_tools == 0:
            return 0.0
            
        success_rate = tool_results["successful_tools"] / total_tools
        
        # Penalize for safety violations
        if tool_results["safety_violations"] > 0:
            success_rate *= 0.7
            
        return round(success_rate, 2)


if __name__ == "__main__":
    # Example usage
    worker = hatchet.worker("ai-agent-worker", workflows=[AIAgentWorkflow()])
    worker.start()

File: examples/use-cases/ai-agents/typescript/ai-agent.ts
/**
 * AI Agent with Streaming and Tool Orchestration
 * 
 * This example demonstrates:
 * - Complex tool call orchestration with timeouts
 * - Conversation state management
 * - Built-in streaming capabilities 
 * - Human-in-the-Loop signaling
 */

import { Hatchet, Context } from '@hatchet-dev/typescript-sdk/v1';

interface CustomerQuery {
  queryId: string;
  customerId: string;
  message: string;
  priority: 'urgent' | 'high' | 'normal' | 'low';
  context: Record<string, any>;
}

interface ToolCall {
  name: string;
  parameters: Record<string, any>;
  timeout: number;
}

interface ConversationState {
  queryId: string;
  messages: Array<{ role: string; content: string; timestamp: number }>;
  toolResults: Array<{ tool: string; success: boolean; result: any; duration: number }>;
  safetyFlags: string[];
  humanReviewRequired: boolean;
}

const hatchet = new Hatchet();

export default hatchet.workflow('ai-agent-workflow', async (context: Context) => {
  return {
    // Step 1: Analyze and route the query
    analyzeQuery: hatchet.step('analyze-query', {
      timeout: '30s',
      retries: 2
    })(async (context: Context): Promise<{
      routing: any;
      toolPlan: ToolCall[];
      riskAssessment: any;
    }> => {
      const query = context.workflowInput() as CustomerQuery;
      
      console.log(`Analyzing query ${query.queryId} from customer ${query.customerId}`);
      
      // Intelligent query analysis
      const routing = analyzeQueryIntent(query);
      const toolPlan = planToolExecution(query);
      const riskAssessment = assessSecurityRisk(query);
      
      return {
        routing,
        toolPlan, 
        riskAssessment
      };
    }),

    // Step 2: Execute tools with orchestration
    executeTools: hatchet.step('execute-tools', {
      timeout: '120s',
      retries: 1,
      parents: ['analyze-query']
    })(async (context: Context): Promise<ConversationState> => {
      const { toolPlan, riskAssessment } = context.stepOutput('analyze-query');
      const query = context.workflowInput() as CustomerQuery;
      
      const state: ConversationState = {
        queryId: query.queryId,
        messages: [{
          role: 'user',
          content: query.message,
          timestamp: Date.now()
        }],
        toolResults: [],
        safetyFlags: [],
        humanReviewRequired: false
      };

      // Execute tools with safety constraints
      for (const toolCall of toolPlan) {
        try {
          // Safety check
          if (!checkToolSafety(toolCall, riskAssessment)) {
            state.safetyFlags.push(`Blocked ${toolCall.name} due to safety constraints`);
            continue;
          }

          const startTime = Date.now();
          
          // Execute with timeout
          const result = await executeToolWithTimeout(toolCall, context);
          const duration = Date.now() - startTime;
          
          state.toolResults.push({
            tool: toolCall.name,
            success: true,
            result,
            duration
          });

          console.log(`Tool ${toolCall.name} completed in ${duration}ms`);
          
        } catch (error) {
          state.toolResults.push({
            tool: toolCall.name,
            success: false,
            result: error.message,
            duration: 0
          });
          
          console.error(`Tool ${toolCall.name} failed:`, error.message);
        }
      }

      // Determine if human review is needed
      state.humanReviewRequired = shouldRequireHumanReview(state, riskAssessment);

      return state;
    }),

    // Step 3: Generate response with streaming
    generateResponse: hatchet.step('generate-response', {
      timeout: '60s',
      parents: ['execute-tools']
    })(async (context: Context): Promise<{
      response: any;
      streamingEnabled: boolean;
      humanEscalation?: any;
    }> => {
      const state = context.stepOutput('execute-tools') as ConversationState;
      const { routing } = context.stepOutput('analyze-query');
      
      if (state.humanReviewRequired) {
        // Signal for Human-in-the-Loop
        return {
          response: {
            status: 'escalated',
            queryId: state.queryId,
            reason: 'Complex query requires human expertise',
            summary: generateExecutionSummary(state)
          },
          streamingEnabled: false,
          humanEscalation: {
            priority: routing.priority,
            context: state,
            suggestedActions: generateHumanActions(state)
          }
        };
      }

      // Generate streaming response
      const response = await generateStreamingResponse(state, context);
      
      return {
        response,
        streamingEnabled: true
      };
    })
  };
});

// Helper functions
function analyzeQueryIntent(query: CustomerQuery) {
  const intent = {
    category: categorizeQuery(query.message),
    complexity: assessComplexity(query.message),
    urgency: query.priority,
    customerTier: query.context.customerTier || 'standard'
  };

  return {
    ...intent,
    routingDecision: determineRouting(intent),
    estimatedTime: estimateProcessingTime(intent)
  };
}

function planToolExecution(query: CustomerQuery): ToolCall[] {
  const tools: ToolCall[] = [];
  const message = query.message.toLowerCase();

  // Knowledge base search for documentation queries
  if (message.includes('how to') || message.includes('documentation')) {
    tools.push({
      name: 'knowledge-search',
      parameters: { query: query.message, limit: 5 },
      timeout: 10000
    });
  }

  // System status for health/status queries
  if (message.includes('status') || message.includes('down') || message.includes('error')) {
    tools.push({
      name: 'system-health-check',
      parameters: { services: ['api', 'database', 'cache'] },
      timeout: 15000
    });
  }

  // Account lookup for billing/account queries
  if (message.includes('account') || message.includes('billing')) {
    tools.push({
      name: 'account-lookup',
      parameters: { customerId: query.customerId },
      timeout: 5000
    });
  }

  // Code analysis for technical queries
  if (message.includes('api') || message.includes('integration')) {
    tools.push({
      name: 'code-analyzer',
      parameters: { context: query.context },
      timeout: 20000
    });
  }

  return tools.length > 0 ? tools : [{
    name: 'general-assistant',
    parameters: { query: query.message },
    timeout: 10000
  }];
}

function assessSecurityRisk(query: CustomerQuery) {
  const riskFactors = [];
  const message = query.message.toLowerCase();

  // Check for sensitive operations
  if (message.includes('delete') || message.includes('remove')) {
    riskFactors.push('destructive_operation');
  }

  if (message.includes('admin') || message.includes('root')) {
    riskFactors.push('elevated_privileges');
  }

  // Customer tier risk assessment
  const tierRisk = {
    'enterprise': 'low',
    'pro': 'medium', 
    'starter': 'high'
  }[query.context.customerTier] || 'high';

  return {
    level: riskFactors.length > 0 ? 'high' : tierRisk,
    factors: riskFactors,
    requiresApproval: riskFactors.length > 1
  };
}

function checkToolSafety(toolCall: ToolCall, riskAssessment: any): boolean {
  // High-risk tools require additional checks
  const highRiskTools = ['system-admin', 'database-query', 'user-management'];
  
  if (highRiskTools.includes(toolCall.name) && riskAssessment.level === 'high') {
    return false;
  }

  return true;
}

async function executeToolWithTimeout(toolCall: ToolCall, context: Context): Promise<any> {
  const tools = {
    'knowledge-search': async (params: any) => ({
      results: [
        { title: 'API Documentation', relevance: 0.95 },
        { title: 'Integration Guide', relevance: 0.87 }
      ],
      totalFound: 2
    }),
    
    'system-health-check': async (params: any) => ({
      status: 'operational',
      services: {
        api: { status: 'healthy', responseTime: '45ms' },
        database: { status: 'healthy', connections: 12 },
        cache: { status: 'healthy', hitRate: '94%' }
      }
    }),
    
    'account-lookup': async (params: any) => ({
      customerId: params.customerId,
      tier: 'pro',
      status: 'active',
      usage: { current: '1,250', limit: '10,000' }
    }),
    
    'code-analyzer': async (params: any) => ({
      analysis: 'Code structure looks good',
      suggestions: ['Add error handling', 'Implement rate limiting'],
      score: 85
    }),
    
    'general-assistant': async (params: any) => ({
      response: 'I can help you with that. Let me analyze your request...',
      confidence: 0.9
    })
  };

  const executor = tools[toolCall.name];
  if (!executor) {
    throw new Error(`Unknown tool: ${toolCall.name}`);
  }

  return new Promise(async (resolve, reject) => {
    const timer = setTimeout(() => {
      reject(new Error(`Tool ${toolCall.name} timed out after ${toolCall.timeout}ms`));
    }, toolCall.timeout);

    try {
      const result = await executor(toolCall.parameters);
      clearTimeout(timer);
      resolve(result);
    } catch (error) {
      clearTimeout(timer);
      reject(error);
    }
  });
}

function shouldRequireHumanReview(state: ConversationState, riskAssessment: any): boolean {
  // Require human review if:
  // - High risk assessment
  // - Multiple tool failures
  // - Safety flags raised
  // - Complex query with low confidence
  
  const failedTools = state.toolResults.filter(r => !r.success).length;
  
  return (
    riskAssessment.level === 'high' ||
    failedTools > 1 ||
    state.safetyFlags.length > 0 ||
    riskAssessment.requiresApproval
  );
}

async function generateStreamingResponse(state: ConversationState, context: Context): Promise<any> {
  const successfulResults = state.toolResults.filter(r => r.success);
  
  // Simulate streaming response generation
  const responseChunks = [
    'Based on my analysis',
    'I found the following information',
    'Here are the key points',
    'Let me know if you need clarification'
  ];

  return {
    status: 'completed',
    queryId: state.queryId,
    message: responseChunks.join(' '),
    toolsUsed: successfulResults.length,
    confidence: calculateConfidence(state),
    processingTime: state.toolResults.reduce((sum, r) => sum + r.duration, 0),
    streamingChunks: responseChunks
  };
}

function generateExecutionSummary(state: ConversationState): string {
  const successful = state.toolResults.filter(r => r.success).length;
  const failed = state.toolResults.filter(r => !r.success).length;
  
  return `Executed ${successful + failed} tools (${successful} successful, ${failed} failed). ${state.safetyFlags.length} safety flags raised.`;
}

function generateHumanActions(state: ConversationState): string[] {
  const actions = [];
  
  if (state.toolResults.some(r => !r.success)) {
    actions.push('Review failed tool executions');
  }
  
  if (state.safetyFlags.length > 0) {
    actions.push('Assess safety concerns');
  }
  
  actions.push('Provide expert guidance');
  
  return actions;
}

function categorizeQuery(message: string): string {
  const categories = {
    'technical': ['api', 'integration', 'webhook', 'error', 'bug'],
    'billing': ['billing', 'payment', 'invoice', 'subscription'],
    'general': ['help', 'question', 'how to']
  };

  for (const [category, keywords] of Object.entries(categories)) {
    if (keywords.some(keyword => message.toLowerCase().includes(keyword))) {
      return category;
    }
  }

  return 'general';
}

function assessComplexity(message: string): 'low' | 'medium' | 'high' {
  const wordCount = message.split(' ').length;
  const technicalTerms = ['api', 'webhook', 'authentication', 'database', 'integration'];
  const hasTechnicalTerms = technicalTerms.some(term => message.toLowerCase().includes(term));

  if (wordCount > 100 || hasTechnicalTerms) return 'high';
  if (wordCount > 50) return 'medium';
  return 'low';
}

function determineRouting(intent: any): string {
  if (intent.category === 'technical' && intent.complexity === 'high') {
    return 'senior-technical-agent';
  }
  if (intent.category === 'billing') {
    return 'billing-specialist';
  }
  return 'general-agent';
}

function estimateProcessingTime(intent: any): number {
  const baseTime = 30; // seconds
  const complexityMultiplier = {
    'low': 1,
    'medium': 1.5,
    'high': 2.5
  }[intent.complexity];

  return Math.round(baseTime * complexityMultiplier);
}

function calculateConfidence(state: ConversationState): number {
  const totalTools = state.toolResults.length;
  if (totalTools === 0) return 0;

  const successRate = state.toolResults.filter(r => r.success).length / totalTools;
  
  // Penalize for safety flags
  const safetyPenalty = state.safetyFlags.length * 0.1;
  
  return Math.max(0, Math.round((successRate - safetyPenalty) * 100) / 100);
}